compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -28,37 +29,228 @@ pub enum AccessLevel {
+        match tag {
+            AccessLevel::Public => &mut self.public,
+            AccessLevel::Exported => &mut self.exported,
+            AccessLevel::Reachable => &mut self.reachable,
+            AccessLevel::ReachableFromImplTrait => &mut self.reachable_from_impl_trait,
+        }
+    }
+
+    fn merge<T: DefIdTree>(&mut self, vis: EffectiveVisibility, tag: AccessLevel, tree: T) {
+        let current_effective_vis = self.get_mut(tag);
+        if current_effective_vis.is_public() {
+            return;
+        }
+
+        if vis.is_public() || (vis.is_some() && current_effective_vis.is_none()) {
+            current_effective_vis.vis = vis.vis;
+            return;
+        }
+
+        if  let Some(Visibility::Restricted(restricted_id)) = vis.vis &&
+            let Some(Visibility::Restricted(current_restricted_id)) = current_effective_vis.vis {
+            if tree.is_descendant_of(restricted_id, current_restricted_id) {
+                current_effective_vis.vis = vis.vis;
+            }
+        }
+    }
+
+    fn set<T: DefIdTree>(&mut self, vis: EffectiveVisibility, tag: AccessLevel, tree: T) {
+        self.merge(vis, tag, tree);
+    }
+
+    pub fn update<T: DefIdTree>(&mut self, vis: EffectiveVisibility, tag: AccessLevel, tree: T) {
+        if tag == AccessLevel::ReachableFromImplTrait {
+            self.merge(vis, tag, tree);
+            return;
+        }
+
+        for level in [AccessLevel::Public, AccessLevel::Exported, AccessLevel::Reachable] {
+            if level <= tag {
+                self.merge(vis, level, tree);
+            }
+        }
+    }
+
+    pub fn calculate_effective_vis<T: DefIdTree>(
+        current_vis: Visibility,
+        inherited_effective_vis: EffectiveVisibility,
+        tree: T,
+    ) -> Option<EffectiveVisibility> {
+        match current_vis {
+            Visibility::Public => Some(inherited_effective_vis),
+            Visibility::Restricted(current_restricted_id) => {
+                if inherited_effective_vis.is_public() {
+                    return Some(EffectiveVisibility::from(current_vis));
+                };
+
+                if let Some(Visibility::Restricted(inherited_restricted_id)) =
+                    inherited_effective_vis.vis
+                {
+                    if tree.is_descendant_of(current_restricted_id, inherited_restricted_id) {
+                        return Some(EffectiveVisibility::from(current_vis));
+                    } else {
+                        return Some(EffectiveVisibility::from(
+                            inherited_effective_vis.expect_vis(),
+                        ));
+                    }
+                };
+
+                None
+            }
+        }
+    }
+}
+
 #[derive(Debug, Clone)]
 pub struct AccessLevels<Id = LocalDefId> {
-    pub map: FxHashMap<Id, AccessLevel>,
+    pub access_levels: FxHashMap<Id, AccessLevel>,
+    pub effective_vis: FxHashMap<LocalDefId, EffectiveVisibilityValue>,
```

Comment:
This `access_levels` field can be removed entirely, any methods implementable on `access_levels` are also implementable on `effective_vis`
It's also better to make `effective_vis` private.

That's true even for methods like `insert`, e.g.
```rust
access_levels.insert(def_id, level)
```
is equivalent to something like
```
effective_vis.entry(def_id).or_default().update(EffectiveVisibility::Pub, level, tree)
```